### PR TITLE
PF-469 Add controlled notebook instance create & get endpoints

### PR DESCRIPTION
**Pull request by wchamber**
Add open api definitions and implementation for controlled AI Platform notebook instance resources.
- async create endpoint
- create-result endpoint to use to poll/retrieve results
- get endpoint

TestRunner client tests to follow in another PR. Need to add a deletion endpoint so that the workspace can be deleted without running into PF-612.

### CODE REVIEW

**Comment by wchamber** [comment]
@rtitle FYI initial endpoints.

File: src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java [hunk 1]
```diff
@@ -150,6 +162,118 @@ public ControlledGcpResourceApiController(
...
+  }
+
+  /**
+   * Extract a list of ControlledResourceIamRoles from the common fields of a controlled resource
+   * request body, and validate that it's shaped appropriately for the specified AccessScopeType.
+   *
+   * <p>Shared access resources must not specify private resource roles. Private access resources
+   * must specify at least one private resource role.
+   */
+  private static List<ControlledResourceIamRole> getAndValidatePrivateRoles(
```
Comment (ddietterich) [inline src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java:255]: This looks familiar. Is this a function we should move into a utility?

File: src/main/java/bio/terra/workspace/service/resource/ValidationUtils.java [hunk 2]
```diff
@@ -53,7 +56,28 @@ public static void validateAiNotebookInstanceId(String name) {
...
-          "Invalid AI Notebook instance name specified. Name must be 1 to 62 alphanumeric lower case characters or underscores, where the first letter is a lower case letter.");
+          "Invalid AI Notebook instance name specified. Name must be 1 to 63 alphanumeric lower case characters or underscores, where the first letter is a lower case letter.");
+    }
+  }
+
+  public static void validate(ApiGcpAiNotebookInstanceCreationParameters creationParameters) {
+    validateAiNotebookInstanceId(creationParameters.getInstanceId());
+    // OpenApi one-of fields aren't being generated correctly, so we do manual one-of fields.
+    if ((creationParameters.getVmImage() == null)
+        == (creationParameters.getContainerImage() == null)) {
```
Comment (ddietterich) [inline src/main/java/bio/terra/workspace/service/resource/ValidationUtils.java:67]: Huh. I would have said `&&` but sure.

File: src/main/java/bio/terra/workspace/service/resource/ValidationUtils.java [hunk 3]
```diff
@@ -53,7 +56,28 @@ public static void validateAiNotebookInstanceId(String name) {
...
+    validate(creationParameters.getVmImage());
+  }
+
+  private static void validate(ApiGcpAiNotebookInstanceVmImage vmImage) {
+    if (vmImage == null) {
+      return;
+    }
+    if ((vmImage.getImageName() == null) == (vmImage.getImageFamily() == null)) {
+      throw new InconsistentFieldsException(
+          "Exactly one of imageName or imageFaily must be specified for a valid vmImage.");
```
Comment (ddietterich) [inline src/main/java/bio/terra/workspace/service/resource/ValidationUtils.java:80]: typo: imageFaily

File: src/main/resources/api/service_openapi.yaml [hunk 4]
```diff
@@ -1798,6 +1877,33 @@ components:
         attributes:
           $ref: '#/components/schemas/GcpGcsBucketAttributes'
 
+    GcpAiNotebookInstanceAttributes:
+      description: >-
+        AI Notebook instance properties included in post-creation, get, and update. Others must be
+        retrieved from GCS using the specification.
```
Comment (ddietterich) [inline src/main/resources/api/service_openapi.yaml:1883]: nit: Probably not from GCS

**Review by ddietterich — APPROVED** [review]
LGTM - a couple of comments

File: src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java (hunk 1 above)
Comment (wchamber) [reply src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java]: Zach and I were racing PRS. Using the function he added in his BQ PR. (I had meant to finish pulling this out of `createBucket`).

Comment (zloery) [reply src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java]: The similarity is probably with #307, which also pulled this function out of `createBucket`. there's going to be a small merge conflict here, but I think this controller is approximately the right place for it. maybe we move it to a utility once we support non-GCP controlled resources?

**Review by zloery — APPROVED** [review]

File: src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java (hunk 1 above)
Comment (wchamber) [reply src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java]: Thanks, just merged and updated, PTAL

File: src/main/resources/api/service_openapi.yaml [hunk 5]
```diff
@@ -1811,6 +1917,8 @@ components:
           $ref: '#/components/schemas/DataRepoSnapshotAttributes'
         gcpGcsBucket:
           $ref: '#/components/schemas/GcpGcsBucketAttributes'
+        gcpAiNotebookInstance:
```
[the remark this answers is not in the document]
Comment (wchamber) [reply src/main/resources/api/service_openapi.yaml:1975]: Thanks, I missed that. Adding a TODO, but will update in a follow-up PR.

**Pull request merged**